### PR TITLE
Branch id and article id mapping

### PR DESCRIPTION
**Pull request by kuzmaka**

### CODE REVIEW

File: src/SpecialCreateBranch.php [hunk 1]
```diff
@@ -52,40 +58,47 @@ protected function getFormFields() {
...
+
 		$comment = CommentStoreComment::newUnsavedComment(
 			wfMessage( 'createbranch-revision-comment' )->inContentLanguage()->text()
 		);
 
 		$page->newPageUpdater( $this->getUser() )
-			->setContent( SlotRecord::MAIN, $text )
+			->setContent( SlotRecord::MAIN, ContentHandler::makeContent('Branch root page', $title) )
+			->setContent( 'lakat', $content )
 			->saveRevision( $comment );
```
Comment (leomarlo) [inline src/SpecialCreateBranch.php:97]: How does this local file storage solution work? 
SlotRecord::MAIN is a global variable?

File: src/SpecialCreateBranch.php [hunk 2]
```diff
@@ -78,29 +80,17 @@ public function onSubmit( array $data ) {
...
-		$text .= "|-\n";
-		$text .= sprintf( "| Token || %s\n", $data['Token'] );
-		$text .= "|-\n";
-		$text .= sprintf( "| Type || %s\n", $data['BranchType'] );
-		$text .= "|-\n";
-		$text .= sprintf( "| Allow conflicts || %s\n", $data['AllowConflicts'] ? 'yes' : 'no' );
-		$text .= "|}\n";
-		$content = ContentHandler::makeContent( $text, $title );
+		$json = FormatJson::encode( [ 'BranchId' => $branchId ] + $data );
+		$content = ContentHandler::makeContent( $json, $title, CONTENT_MODEL_JSON );
```
Comment (leomarlo) [inline src/SpecialCreateBranch.php:84]: So the process for storing thing in json here is: First to encode, then to use the ContentHandler and  that will write it to storage?

File: src/LakatArticleMetadata.php [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+			throw new Exception('Article has invalid metadata: articleId field is not set');
+		}
+		return $metadata->articleId;
+	}
+
+	private static function getPageMetadata(WikiPage $page): object {
+		$revisionRecord = $page->getRevisionRecord();
+		if (!$revisionRecord->hasSlot('lakat')) {
+			throw new Exception('Page has no metadata slot');
+		}
```
Comment (leomarlo) [inline src/LakatArticleMetadata.php:51]: So the logic here is that the slot lakat is created for each now branch?

File: src/Storage/LakatStorageStub.php [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+		if ( !@mkdir( $dir, 0755, true ) ) {
+			throw new Exception( 'Failed to create branch directory ' . $dir );
+		}
+
+		// store branch definition
+		if (
+			!file_put_contents(
+				$dir . '/.branch',
+				json_encode( compact( 'name' ) + $options, JSON_THROW_ON_ERROR )
+			)
```
Comment (leomarlo) [inline src/Storage/LakatStorageStub.php:32]: Is this the json-encoded branch-metadata, or just temporary storage?

File: src/Storage/LakatStorageStub.php [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+		if ( $content === false ) {
+			throw new Exception( 'Failed to read article from file ' . $filename );
+		}
+
+		return $content;
+	}
+
+	private function getSlug( string $name ) {
+		return preg_replace( '/[^a-z0-9]/', '_', strtolower( trim( $name ) ) );
+	}
```
Comment (leomarlo) [inline src/Storage/LakatStorageStub.php:87]: What is this function designated for?

**Review by leomarlo — APPROVED** [review]

File: src/LakatArticleMetadata.php (hunk 3 above)
Comment (kuzmaka) [reply src/LakatArticleMetadata.php:51]: Yes, for each branch and for each article in a branch.

File: src/Storage/LakatStorageStub.php (hunk 4 above)
Comment (kuzmaka) [reply src/Storage/LakatStorageStub.php:32]: This is just a temporary stub for the remote storage. I had to simulate it somehow.

File: src/Storage/LakatStorageStub.php (hunk 5 above)
Comment (kuzmaka) [reply src/Storage/LakatStorageStub.php:87]: It's an utility function that converts arbitrary name to restricted set of characters to have nicer filename without  special characters or spaces.

File: src/SpecialCreateBranch.php (hunk 1 above)
Comment (kuzmaka) [reply src/SpecialCreateBranch.php:97]: To clarify, this code is related to metadata stored as part of wiki page content. Just to distinguish from local file storage that I created as quick temporary replacement of remote distributed db.

`SlotRecord::MAIN` is a `SlotRecord `class constant, it's just a string `'main'`. In mediawiki slot `main` is dedicated to the main wiki page content. Main slot is required for any wiki page.

**Pull request merged**